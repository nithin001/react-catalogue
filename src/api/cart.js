import instance from './base';
import Ajv from 'ajv';

export const get = () => {
  return new Promise((resolve, reject) => {
    instance.get('/cart')
      .then((response) => {
        if (response.data === undefined) {
          reject('backend_error');
          return;
        }
        const data = response.data;
        var ajv = new Ajv();
        var validate = ajv.compile(get_schema);
        var valid = validate(data);
        if (valid) {
          resolve(data);
        } else {
          reject('backend_error');
        }
      })
      .catch(error => {
        reject('backend_error');
      });
  });
};

export const put = (request) => {
  return new Promise((resolve, reject) => {
    instance.put('/cart', { data: request })
      .then((response) => {
        if (response.data === undefined) {
          reject('backend_error');
          return;
        }
        const data = response.data;
        var ajv = new Ajv();
        var validate = ajv.compile(put_schema);
        var valid = validate(data);
        if (valid) {
          resolve(data);
        } else {
          reject('backend_error');
        }
      })
      .catch(error => {
        reject('backend_error');
      });
  });
};

const get_schema = {
  '$id': 'http://challenge.monoqi.net/cart',
  'type': 'object',
  'definitions': {},
  '$schema': 'http://json-schema.org/draft-07/schema#',
  'properties': {
    'lines': {
      '$id': '/properties/lines',
      'type': 'array',
      'items': {
        '$id': '/properties/lines/items',
        'type': 'object',
        'properties': {
          'sku': {
            '$id': '/properties/lines/items/properties/sku',
            'type': 'string',
            'title': 'The Sku Schema ',
            'default': '',
            'examples': [
              '199203'
            ]
          },
          'quantity': {
            '$id': '/properties/lines/items/properties/quantity',
            'type': 'integer',
            'title': 'The Quantity Schema ',
            'default': 0,
            'examples': [
              3
            ]
          }
        },
        'required': ['sku', 'quantity']
      }
    }
  },
  'required': ['lines']
};

const put_schema = {
  '$id': 'http://example.com/example.json',
  'type': 'object',
  'definitions': {},
  '$schema': 'http://json-schema.org/draft-07/schema#',
  'properties': {
    'lines': {
      '$id': '/properties/lines',
      'type': 'array',
      'items': {
        '$id': '/properties/lines/items',
        'type': 'object',
        'properties': {
          'sku': {
            '$id': '/properties/lines/items/properties/sku',
            'type': 'string',
            'title': 'The Sku Schema ',
            'default': '',
            'examples': [
              '199203'
            ]
          },
          'name': {
            '$id': '/properties/lines/items/properties/name',
            'type': 'string',
            'title': 'The Name Schema ',
            'default': '',
            'examples': [
              'IPad'
            ]
          },
          'price': {
            '$id': '/properties/lines/items/properties/price',
            'type': 'object',
            'properties': {
              'amount': {
                '$id': '/properties/lines/items/properties/price/properties/amount',
                'type': 'integer',
                'title': 'The Amount Schema ',
                'default': 0,
                'examples': [
                  229
                ]
              },
              'currency': {
                '$id': '/properties/lines/items/properties/price/properties/currency',
                'type': 'string',
                'title': 'The Currency Schema ',
                'default': '',
                'examples': [
                  'EUR'
                ]
              }
            },
            'required': [
              'amount',
              'currency'
            ]
          },
          'quantity': {
            '$id': '/properties/lines/items/properties/quantity',
            'type': 'integer',
            'title': 'The Quantity Schema ',
            'default': 0,
            'examples': [
              3
            ]
          },
          'lineTotal': {
            '$id': '/properties/lines/items/properties/lineTotal',
            'type': 'object',
            'properties': {
              'amount': {
                '$id': '/properties/lines/items/properties/lineTotal/properties/amount',
                'type': 'integer',
                'title': 'The Amount Schema ',
                'default': 0,
                'examples': [
                  687
                ]
              },
              'currency': {
                '$id': '/properties/lines/items/properties/lineTotal/properties/currency',
                'type': 'string',
                'title': 'The Currency Schema ',
                'default': '',
                'examples': [
                  'EUR'
                ]
              }
            },
            'required': [
              'amount',
              'currency'
            ]
          }
        },
        'required': [
          'sku',
          'name',
          'price',
          'quantity',
          'lineTotal'
        ]
      }
    },
    'total': {
      '$id': '/properties/total',
      'type': 'object',
      'properties': {
        'amount': {
          '$id': '/properties/total/properties/amount',
          'type': 'integer',
          'title': 'The Amount Schema ',
          'default': 0,
          'examples': [
            1126
          ]
        },
        'currency': {
          '$id': '/properties/total/properties/currency',
          'type': 'string',
          'title': 'The Currency Schema ',
          'default': '',
          'examples': [
            'EUR'
          ]
        }
      },
      'required': [
        'amount',
        'currency'
      ]
    }
  },
  'required': [
    'lines',
    'total'
  ]
};