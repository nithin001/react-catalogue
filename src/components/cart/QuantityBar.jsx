import React from 'react';
export const QuantityBar = ({ id, quantity, addToCart, removeFromCart }) => {
  return <div className={'quantity-bar'}>
    <button className={'add-to-cart'} onClick={() => {addToCart(id);}}>Add to cart</button>
    <span className={'quantity'}>{quantity}</span>
    <button className={'remove-from-cart'} onClick={() => {removeFromCart(id);}}>Remove from cart</button>
  </div>;
};