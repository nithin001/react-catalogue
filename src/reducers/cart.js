import Immutable from 'immutable';
const cart = (state = Immutable.Map(), action) => {
  switch (action.type) {
    case 'POPULATE_CART': {
      const items = action.payload.lines;
      return items.reduce((result, item) => {
        return result.set(item.sku, item.quantity);
      }, Immutable.Map());
    }
    case 'ADD_TO_CART': {
      const sku = action.payload;
      if (state.has(sku)) {
        const quantity = state.get(sku);
        return state.set(sku, quantity + 1);
      }
      return state.set(sku, 1);
    }
    case 'REMOVE_FROM_CART': {
      const sku = action.payload;
      const quantity = state.get(sku);
      if (quantity === 1) {
        return state.delete(sku);
      }
      return state.set(sku, quantity - 1);
    }
  }
  return state;
};
export default cart;