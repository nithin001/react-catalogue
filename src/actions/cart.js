import { get } from '../api/cart';
export const load = () => {
  return (dispatch) => {
    dispatch({ type: 'LOAD_CART_START' });
    return get()
      .then((cart) => {
        dispatch({ type: 'POPULATE_CART', payload: cart });
        dispatch({ type: 'LOAD_CART_COMPLETE' });
      })
      .catch(() => {
        dispatch({ type: 'LOAD_ERROR' });
      });
  };
};