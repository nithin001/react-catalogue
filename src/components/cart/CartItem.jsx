import React from 'react';
import { connect } from 'react-redux';
import { QuantityBar } from './QuantityBar';
import { addToCart, removeFromCart } from '../../actions/cart';
export const CartItem = ({ sku, price, name, quantity, addToCart, removeFromCart }) => {
  return <div className={'cart-item'}>
    <span className={'name'}>{name}</span>
    <span className={'price-amount'}>{price.amount}</span>
    <span className={'price-currency'}>{price.currency}</span>
    <QuantityBar sku={sku} quantity={quantity} addToCart={addToCart} removeFromCart={removeFromCart}/>
  </div>;
};

export const mapStateToProps = (state, ownProps) => {
  const article = state.catalog.filter(item => {
    return item.get('sku') === ownProps.sku;
  }).first();
  const price = {
    amount: article.getIn(['price', 'amount']),
    currency: article.getIn(['price', 'currency']),
  };
  return { price, name: article.get('name') };
};

export const mapDispatchToProp = (dispatch) => ({
    addToCart: (sku) => dispatch(addToCart(sku)),
    removeFromCart: (sku) => dispatch(removeFromCart(sku)),
  }
);

export default connect(mapStateToProps, mapDispatchToProp)(CartItem);
