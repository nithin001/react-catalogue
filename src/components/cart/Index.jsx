import React from 'react';
import { connect } from 'react-redux';
import CartItem from './CartItem';
export const Cart = ({ cart }) => {
  const skus = [];
  cart.keySeq().forEach(key => skus.push(key));
  const items = skus.map((key) => (<CartItem
    key={key}
    sku={key}
    quantity={cart.get(key)}
  />));
  return <div className={'cart'}>{items}</div>;
};

export const mapStateToProps = (state) => {
  return { cart: state.cart };
};

export default connect(mapStateToProps, null)(Cart);
