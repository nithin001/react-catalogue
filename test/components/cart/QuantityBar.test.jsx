import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { QuantityBar } from '../../../src/components/cart/QuantityBar';
configure({ adapter: new Adapter() });

describe('QuantityBar component', () => {
  it('should render the QuantityBar component with current count and two buttons', () => {
    const addToCart = jest.fn();
    const removeFromCart = jest.fn();
    const props = { id: '199203', quantity: 1, addToCart, removeFromCart };
    const quantityBar = renderer.create(<QuantityBar
      {...props}
    />);
    expect(quantityBar).toMatchSnapshot();
  });

  it('should call add to cart function on click of add to cart', () => {
    const addToCart = jest.fn();
    const removeFromCart = jest.fn();
    const props = { id: '199203', quantity: 1, addToCart, removeFromCart };
    const quantityBar = shallow(<QuantityBar
      {...props}
    />);
    quantityBar.find('button').at(0).simulate('click');
    expect(addToCart.mock.calls.length).toBe(1);
    expect(addToCart.mock.calls[0][0]).toEqual('199203');
  });

  it('should call remove from cart function on click of remove from cart', () => {
    const addToCart = jest.fn();
    const removeFromCart = jest.fn();
    const props = { id: '199203', quantity: 1, addToCart, removeFromCart };
    const quantityBar = shallow(<QuantityBar
      {...props}
    />);
    quantityBar.find('button').at(1).simulate('click');
    expect(removeFromCart.mock.calls.length).toBe(1);
    expect(removeFromCart.mock.calls[0][0]).toEqual('199203');
  });
});
