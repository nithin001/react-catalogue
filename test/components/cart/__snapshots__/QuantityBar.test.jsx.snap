// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityBar component should render the QuantityBar component with current count and two buttons 1`] = `
<div
  className="quantity-bar"
>
  <button
    className="add-to-cart"
    onClick={[Function]}
  >
    Add to cart
  </button>
  <span
    className="quantity"
  >
    1
  </span>
  <button
    className="remove-from-cart"
    onClick={[Function]}
  >
    Remove from cart
  </button>
</div>
`;
